### PR TITLE
ERC-1155 offers

### PR DESCRIPTION
**Pull request by orionstardust**

### CODE REVIEW

File: contracts/Aavegotchi/facets/ERC1155BuyOrderFacet.sol [hunk 1]
```diff
@@ -0,0 +1,256 @@
...
+
+    function placeERC1155BuyOrder(
+        address _erc1155TokenAddress,
+        uint256 _erc1155TokenId,
+        uint256 _priceInWei,
+        uint256 _quantity,
+        uint256 _duration
+    ) external {
+        uint256 cost = _quantity * _priceInWei;
+        require(cost >= 1e18, "ERC1155BuyOrder: cost should be 1 GHST or larger");
```
Comment (cinnabarhorse) [inline contracts/Aavegotchi/facets/ERC1155BuyOrderFacet.sol:74]: Let's change this to 0.001 GHST @orionstardust

Comment (orionstardust) [reply contracts/Aavegotchi/facets/ERC1155BuyOrderFacet.sol:74]: Updated.

File: contracts/Aavegotchi/facets/ERC1155BuyOrderFacet.sol [hunk 2]
```diff
@@ -0,0 +1,256 @@
...
+        }
+    }
+
+    function cancelERC1155BuyOrder(uint256 _buyOrderId) external {
+        address sender = LibMeta.msgSender();
+        ERC1155BuyOrder memory erc1155BuyOrder = s.erc1155BuyOrders[_buyOrderId];
+        require(erc1155BuyOrder.timeCreated != 0, "ERC1155BuyOrder: ERC1155 buyOrder does not exist");
+        require(sender == erc1155BuyOrder.buyer, "ERC1155BuyOrder: Only buyer can call this function");
+        require((erc1155BuyOrder.cancelled == false) && (erc1155BuyOrder.completed == false), "ERC1155BuyOrder: Already processed");
+        if (erc1155BuyOrder.duration > 0) {
```
[the remark this answers is not in the document]
Comment (orionstardust) [reply contracts/Aavegotchi/facets/ERC1155BuyOrderFacet.sol:147]: Yes, expired and canceled are different state.

[the remark this answers is not in the document]
Comment (cinnabarhorse) [reply contracts/Aavegotchi/facets/ERC1155BuyOrderFacet.sol:147]: So why can only the buyer cancel it? 
`require(sender == erc1155BuyOrder.buyer, "ERC1155BuyOrder: Only buyer can call this function");`

File: contracts/Aavegotchi/facets/ERC1155BuyOrderFacet.sol [hunk 3]
```diff
@@ -0,0 +1,264 @@
...
+            if ((erc1155BuyOrder.duration == 0) || (erc1155BuyOrder.timeCreated + erc1155BuyOrder.duration >= block.timestamp)) {
+                // Transfer GHST
+                uint256 oldCost = erc1155BuyOrder.quantity * erc1155BuyOrder.priceInWei;
+                uint256 difference = cost - oldCost;
+                if (cost > oldCost) {
+                    require(ghstBalance >= difference, "ERC1155BuyOrder: Not enough GHST!");
+
+                    //Transfer the difference to the contract
+                    LibERC20.transferFrom(s.ghstContract, sender, address(this), difference);
+                } else if (cost < oldCost) {
```
Comment (cinnabarhorse) [inline contracts/Aavegotchi/facets/ERC1155BuyOrderFacet.sol:98]: What will happen if cost == oldCost? @orionstardust Did you explore this case?

File: contracts/Aavegotchi/facets/ERC1155BuyOrderFacet.sol [hunk 4]
```diff
@@ -0,0 +1,264 @@
...
+    }
+
+    function cancelERC1155BuyOrder(uint256 _buyOrderId) external {
+        address sender = LibMeta.msgSender();
+        ERC1155BuyOrder memory erc1155BuyOrder = s.erc1155BuyOrders[_buyOrderId];
+        require(erc1155BuyOrder.timeCreated != 0, "ERC1155BuyOrder: ERC1155 buyOrder does not exist");
+        require(sender == erc1155BuyOrder.buyer, "ERC1155BuyOrder: Only buyer can call this function");
+        require((erc1155BuyOrder.cancelled == false) && (erc1155BuyOrder.completed == false), "ERC1155BuyOrder: Already processed");
+        if (erc1155BuyOrder.duration > 0) {
+            require(erc1155BuyOrder.timeCreated + erc1155BuyOrder.duration >= block.timestamp, "ERC1155BuyOrder: Already expired");
```
Comment (cinnabarhorse) [inline contracts/Aavegotchi/facets/ERC1155BuyOrderFacet.sol:156]: Why do we have to wait until the offer has expired before cancelling it? @orionstardust 

As a buyer, I should be able to cancel it whenever I want

File: contracts/Aavegotchi/facets/ERC1155BuyOrderFacet.sol [hunk 5]
```diff
@@ -0,0 +1,264 @@
...
+            LibItems.removeFromOwner(sender, erc1155BuyOrder.erc1155TokenId, _quantity);
+            LibItems.addToOwner(erc1155BuyOrder.buyer, erc1155BuyOrder.erc1155TokenId, _quantity);
+            IEventHandlerFacet(s.wearableDiamond).emitTransferSingleEvent(
+                address(this),
+                sender,
+                erc1155BuyOrder.buyer,
+                erc1155BuyOrder.erc1155TokenId,
+                _quantity
+            );
+            LibERC1155.onERC1155Received(address(this), sender, erc1155BuyOrder.buyer, erc1155BuyOrder.erc1155TokenId, _quantity, "");
```
Comment (cinnabarhorse) [inline contracts/Aavegotchi/facets/ERC1155BuyOrderFacet.sol:241]: I think you also need to handle the baazaar hook to delist the item from the baazaar, if it was listed @orionstardust

File: contracts/Aavegotchi/facets/ERC1155BuyOrderFacet.sol (hunk 2 above)
[the remark this answers is not in the document]
Comment (orionstardust) [reply contracts/Aavegotchi/facets/ERC1155BuyOrderFacet.sol:159]: @cinnabarhorse Regarding this, who can cancel offer? seller(erc1155 owner)?
is there any need to cancel offer by owner?
Also, ERC721 buy order is cancelable by buyer only.

File: contracts/Aavegotchi/facets/ERC1155BuyOrderFacet.sol (hunk 3 above)
Comment (orionstardust) [reply contracts/Aavegotchi/facets/ERC1155BuyOrderFacet.sol:99]: @cinnabarhorse if cost and oldCost are same, there's no GHST transfer from price and quantity change. Only buy order updated.

Comment (orionstardust) [reply contracts/Aavegotchi/facets/ERC1155BuyOrderFacet.sol:99]: even if price and quantity changed, cost and oldCost could be same.
For example, old price was 0.1 and quantity was 10 (old cost 1), and now price is 1 and quantity is 1 (cost is 1),
in this case both cost and old cost could be same.

File: contracts/Aavegotchi/facets/ERC1155BuyOrderFacet.sol (hunk 4 above)
Comment (orionstardust) [reply contracts/Aavegotchi/facets/ERC1155BuyOrderFacet.sol:160]: @cinnabarhorse this logic is for checking expired buy orders.
If buy order is already expired, no need to cancel it.

File: contracts/Aavegotchi/facets/ERC1155BuyOrderFacet.sol (hunk 5 above)
Comment (orionstardust) [reply contracts/Aavegotchi/facets/ERC1155BuyOrderFacet.sol:218]: Updated.

File: contracts/Aavegotchi/facets/ERC1155BuyOrderFacet.sol (hunk 2 above)
[the remark this answers is not in the document]
Comment (cinnabarhorse) [reply contracts/Aavegotchi/facets/ERC1155BuyOrderFacet.sol:132]: > @cinnabarhorse Regarding this, who can cancel offer? seller(erc1155 owner)? is there any need to cancel offer by owner? Also, ERC721 buy order is cancelable by buyer only.

If the offer has expired, then maybe anyone should be able to cancel it. What do you think?

File: contracts/Aavegotchi/facets/ERC1155BuyOrderFacet.sol (hunk 4 above)
Comment (cinnabarhorse) [reply contracts/Aavegotchi/facets/ERC1155BuyOrderFacet.sol:133]: > @cinnabarhorse this logic is for checking expired buy orders. If buy order is already expired, no need to cancel it.

Don't you need to cancel the order to get your GHST back?

Comment (cinnabarhorse) [reply contracts/Aavegotchi/facets/ERC1155BuyOrderFacet.sol:133]: Other than cancelling, how do I get my GHST back from an expired order? @orionstardust

**Pull request merged**